# TITANIC PROJECT

This project is about to predict if the passengers of the Titanic are going to survive or not. It is a classification problem.

## Environment

In [1]:
##### DATESET #####
import pandas as pd
import numpy as np

##### VISUALIZATION #####
import matplotlib as mpl
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

##### NOTEBOOK SETTINGS #####
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

## Import and  Data extraction

### Import Dataset

The fist step is to import the data and create a dataset with the train data of the passengers of the titanic.

In [2]:
df = pd.read_csv('Data/train.csv')

## Preprocess of the Dataset

### Problem Study

The **problem** we are going to study in this project is the prediction on the survival or not of a given number of passengers, using the peronal and travel information of the passenger. <br>

The taget Variable is **Survived**.<br>

The rest of the variables we will use for the prediction are the Id, the Class, the Name, Sex, Age of the passenger, the number of Siblings and Spouses, the nomber of Parents and childs, the ticket number, the Fare, the cabin number and the port of embark.

### Analysis of the DataSet

We check the head and tail of the dataset to start the visual and general esploration of the data.

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

We check the shape of the dataset, to know the number of rows and columns.

In [5]:
df.shape

(891, 12)

We have 891 rows and 12 columns

We check the type for each columns.

In [6]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


We have 12 variables, divided on the following types:<br>
    * 7 numerical
    * 5 categorical
But some of the numerical dtypes really are categorical.

### Change of Format of the vars

Eventhough the dtypes gives us the types of the columns, the Variables really have different categories.

#### PassengerId: int64

The **PassengerId** is a number type, the var is a **categorical:nominal** var that gives a unique identificator to each passenger.<br>
We will keep the dtype:int46<br>
Then we will make the index of the dataset in the last step of the Preprocessing.

#### Survived: int64


In [8]:
df.Survived.unique()

array([0, 1])

The **Survived** is a number type, the var is a **categorical:nominal:boolean** var that is a 1 if the passenger survived and a 0 in the passenger deceased.<br>
We will keep the dtype:int64.<br>
**This is the TARGET var**

#### Pclass: int64

In [9]:
df.Pclass.unique()

array([3, 1, 2])

The **Pclass** is a number type, but really is a **categorical:ordinal** var 1 is Fist, 2 is Seccond and 3 is Third.<br>
We will keep the dtype:int64.<br>
In future steps we will indicate that 1 is better than 2 and 3.

#### Name: object

The **Name** is an object type, it is a string, so a **categorical:ordinal** with the name of the passenger<br>
We will keep the dtype:object.<br>

#### Sex: object

In [10]:
df.Sex.unique()

array(['male', 'female'], dtype=object)

The **Sex** is an onject type, is a **categorical:ordinal** for male and females.
We will keep the dtype:object<br>


#### Age: float64

In [11]:
df.Age.unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

The **Age** is an float type, is a **Numerical:continous** of the Age.<br>
I could seem a priori that we can conver to integuer, but we have decimal ages in the values.<br>
We will keep the dtype:float64<br>

#### SibSp: int64

In [12]:
df.SibSp.unique()

array([1, 0, 3, 4, 2, 5, 8])

The **SibSp** is an integer type, is a **Numerical:discreate** of the number of Sibling and Spouses.<br>
We will keep the dtype:int64<br>

#### Parch: int64

In [13]:
df.Parch.unique()

array([0, 1, 2, 5, 3, 4, 6])

The **Parch** is an integer type, is a **Numerical:discreate** of the number of Parent and Children.<br>
We will keep the dtype:int64<br>

#### Ticket: object

The **Ticket** is an object type, is a **Cateforical:nominal** var with the Ticket number<br>
We will keep the dtype:object<br>

#### Fare: float64

The **Fare** is an float type, is a **Numerical:continous** of the fare payed by passenger.<br>
We will keep the dtype:float64<br>

#### Cabin: object

The **Cabin** is an object type, is a **Categorical:nominal** of the number of Cabin in the Titanic.<br>
We will keep the dtype:object<br>

#### Embarked: object

In [14]:
df.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

The **Embarked** is an object type, is a **Categorical:nominal** of the port of embark of the passenger.<br>
We will keep the dtype:object<br>

#### Wrapping up: Variables Types

The summary of the types of variables that we have in this dataset are:

**NUMERICAL**<br>
* Discreate: SibSp,Parch<br>
* Coninous: Fare, Age<br>

**CATEGORICAL**<br>
* Nominal: Embarked, PassengerId, Survived(target), Name, Sex, Ticket, Cabin<br>
* Ordinal: Pclass<br>

### Nans Treatment

We check the number of Nulls for each column:

In [15]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

There are 3 columns with Nans: <br>
    * Age: 177
    * Cabin: 687
    * Embarked: 2

#### Age

In the column Age we found 177 Nans

In [16]:
df.Age.unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

In [17]:
print(f'The percentage of Nans in the Age column is: {round((df.Age.isnull().sum()/df.shape[0])*100,2)}%')


The percentage of Nans in the Age column is: 19.87%


Enthough the number is close to the 20%, we will try to infer the age of the passenger in the EDA step of the project

#### Cabin

In [18]:
print(f'The percentage of Nans in the Cabin column is: {round((df.Cabin.isnull().sum()/df.shape[0])*100,2)}%')


The percentage of Nans in the Cabin column is: 77.1%


In [19]:
df.Cabin.unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

The values of the Cabin as they are do not give us information, but we will keep the column to create a new column in Feature Construction.

#### Embarked

Embarked has only 2 missing values, that we will try to search and complete during the EDA, as the embarked port is a data about the passenger easy to find on the internet as there are only 3 possibilities.

In [20]:
df.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

### DROP COLUMNS AND ROWS

#### DROP COLUMNS

In [21]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

We will only delete the **PassengerId** after using the column to be the index of the Dataset.

In [22]:
df.set_index('PassengerId', inplace=True)

In [23]:
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

#### DROP DUPLICATE ROWS

We check for duplicate rows.

In [24]:
df.duplicated().sum()

0

We do not find any duplicate row.

### Export Dataset for the EDA

In [25]:
df.to_csv('Data/df2.csv', index=False)